runtime/src/iree/base/loop_emscripten.c

```diff
@@ -0,0 +1,101 @@
+                                       void* user_data, iree_loop_t loop);
+
+//===----------------------------------------------------------------------===//
+// iree_loop_emscripten_t
+//===----------------------------------------------------------------------===//
+
+typedef struct iree_loop_emscripten_t {
+  iree_allocator_t allocator;
+
+  // TODO(scotttodd): handle to a "scope"/object (managed in JS), so multiple
+  //                  loops can exist at once
+} iree_loop_emscripten_t;
+
+IREE_API_EXPORT iree_status_t iree_loop_emscripten_allocate(
+    iree_allocator_t allocator, iree_loop_emscripten_t** out_loop_emscripten) {
+  IREE_ASSERT_ARGUMENT(out_loop_emscripten);
+
+  const iree_host_size_t loop_emscripten_size =
+      iree_host_align(sizeof(iree_loop_emscripten_t), iree_max_align_t);
+
+  uint8_t* storage = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_allocator_malloc(allocator, loop_emscripten_size, (void**)&storage));
+  iree_loop_emscripten_t* loop_emscripten = (iree_loop_emscripten_t*)storage;
+  loop_emscripten->allocator = allocator;
+
+  *out_loop_emscripten = loop_emscripten;
+
+  return iree_ok_status();
+}
+
+IREE_API_EXPORT void iree_loop_emscripten_free(
+    iree_loop_emscripten_t* loop_emscripten) {
+  IREE_ASSERT_ARGUMENT(loop_emscripten);
+  iree_allocator_t allocator = loop_emscripten->allocator;
+
+  // TODO(scotttodd): cleanup:
+  //     abort pending operations (neuter callbacks/Promises)
+  //     assert if any work is still outstanding
+
+  // After all operations are cleared we can release the data structures.
+  iree_allocator_free(allocator, loop_emscripten);
+}
+
+static iree_status_t iree_loop_emscripten_run_call(
+    iree_loop_emscripten_t* loop_emscripten, iree_loop_call_params_t* params) {
+  iree_loop_t loop = iree_loop_emscripten(loop_emscripten);
+  return loop_command_call(params->callback.fn, params->callback.user_data,
+                           loop);
+}
+
+// Control function for the Emscripten loop.
+IREE_API_EXPORT iree_status_t
+iree_loop_emscripten_ctl(void* self, iree_loop_command_t command,
+                         const void* params, void** inout_ptr) {
+  IREE_ASSERT_ARGUMENT(self);
+
+  iree_loop_emscripten_t* loop_emscripten = (iree_loop_emscripten_t*)self;
+
+  // NOTE: we return immediately to make this all (hopefully) tail calls.
+  switch (command) {
+    case IREE_LOOP_COMMAND_CALL:
+      return iree_loop_emscripten_run_call(loop_emscripten,
+                                           (iree_loop_call_params_t*)params);
+    case IREE_LOOP_COMMAND_DISPATCH:
+    case IREE_LOOP_COMMAND_WAIT_UNTIL:
+    case IREE_LOOP_COMMAND_WAIT_ONE:
+    case IREE_LOOP_COMMAND_WAIT_ALL:
+    case IREE_LOOP_COMMAND_WAIT_ANY:
+      // TODO(scotttodd): implement these commands
+      return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                              "unimplemented loop command");
+    case IREE_LOOP_COMMAND_DRAIN:
+      return iree_make_status(IREE_STATUS_DEADLINE_EXCEEDED,
```

Comment:
DEADLINE_EXCEEDED?

Reply:
(actually this probably makes sense)

Reply:
Yep, just following what loop.h says implementations should do: https://github.com/iree-org/iree/blob/20048f775f0043f6b8f38441771f6742dec0c9ad/runtime/src/iree/base/loop.h#L256-L263

Reply:
who wrote that? jeez >_>